### PR TITLE
refactor location abstractions

### PR DESCRIPTION
**Pull request by tomerd**
motivation: reduce dependency on string location information, in support of registry based dependencies

changes:
* add location information to PackageReference::Kind
* seperare out sourceControl dependencies to local and remote, where remote carries AbsolutePath and rmeote carries URL
* adjust call sites
* adjust tests and test utilities

rdar://82954076

### CODE REVIEW

File: Sources/PackageModel/Manifest/PackageDependencyDescription.swift
Comment (tomerd) [inline Sources/PackageModel/Manifest/PackageDependencyDescription.swift:44]: 👀  this is another key change

File: Sources/PackageModel/PackageReference.swift
Comment (tomerd) [inline Sources/PackageModel/PackageReference.swift:32]: 👀  this is the key change

File: Sources/SourceControl/Repository.swift
Comment (tomerd) [inline Sources/SourceControl/Repository.swift:70]: 👀  this new enum is another key change

**Comment by tomerd** [comment]
@mattt this is the second PR designed to help with landing SE-0292, I think I prefer to take this one even before #3698

wdyt?

**Comment by tomerd** [comment]
@abertelrud @neonichu @elsh @yim-lee this PR includes refactoring we discussed several times with regards to how location is managed, and that is required to land SE-0292. its actually a pretty small change in the model itself but it has huge impact on call sites, especially in tests. given the large number of the lines changed, please focus on the core model changes which are highlighted in comments

**Comment by tomerd** [comment]
@swift-ci please smoke test

**Comment by tomerd** [comment]
@swift-ci please smoke test

File: Sources/PackageGraph/RepositoryPackageContainer.swift
Comment (tomerd) [inline Sources/PackageGraph/RepositoryPackageContainer.swift:403]: @mattt here you would need to add a case for the registry, and probably we can rename RepositoryPackageContainerProvider to something more generic that deals with all PackageContainer types, perhaps rename to DefaultPackageContainerProvider

File: Sources/PackageLoading/ManifestJSONParser.swift
Comment (tomerd) [inline Sources/PackageLoading/ManifestJSONParser.swift:220]: with the new model, we can do things here much more safely

File: Sources/PackageModel/PackageReference.swift
Comment (tomerd) [inline Sources/PackageModel/PackageReference.swift:70]: after landing this PR, there will be follow up to clean up call sites that still depend on location string. there are very few now, mostly for display so we are getting there

File: Sources/SourceControl/Repository.swift
Comment (tomerd) [inline Sources/SourceControl/Repository.swift:17]: this is for follow up PR

File: Sources/SourceControl/RepositoryManager.swift
Comment (tomerd) [inline Sources/SourceControl/RepositoryManager.swift:553]: this is for follow up PR, its not too bad

File: Sources/PackageModel/PackageIdentity.swift
Comment (yim-lee) [inline Sources/PackageModel/PackageIdentity.swift:49]:
```suggestion
    /// - Parameter urlString: The package's URL.
```

File: Sources/PackageModel/PackageReference.swift
Comment (yim-lee) [reply Sources/PackageModel/PackageReference.swift:32]: What's the difference between `fileSystem` and `localSourceControl`? Do we do anything to ensure `remoteSourceControl` is indeed remote (e.g., what if URL scheme is `file://`?), or it doesn't matter?

**Comment by tomerd** [comment]
> What's the difference between fileSystem and localSourceControl? Do we do anything to ensure remoteSourceControl is indeed remote (e.g., what if URL scheme is file://?), or it doesn't matter?

this is exactly one of the goals of this PR, `fileSystem` is a local dependency that is not necessarily a git repo, while localSourceControl must be a git repo. with this distinction the downstream code is much safe. we do not validate the prefix for URLs put into remoteSourceControl, that is a good idea to add

**Comment by tomerd** [comment]
@swift-ci please smoke test

**Comment by tomerd** [comment]
@swift-ci please test package compatibility

File: Sources/PackageModel/PackageReference.swift
Comment (mattt) [reply Sources/PackageModel/PackageReference.swift:32]: Is there an opportunity for us to reuse the `SourceControl.Location` enumeration here?

https://github.com/apple/swift-package-manager/blob/51fe9f947e18fe3daccf3c2aaf6391136ba2b801/Sources/PackageModel/Manifest/PackageDependencyDescription.swift#L41-L44

**Comment by mattt** [comment]
> @mattt this is the second PR designed to help with landing SE-0292, I think I prefer to take this one even before #3698. wdyt?

If #3698 is ready to go, I think it'd be great to lock in our gains with that. But you're closer to the source material and, so I defer to you on the best order of operations.

File: Sources/PackageModel/PackageReference.swift
Comment (tomerd) [reply Sources/PackageModel/PackageReference.swift:32]: yes, my plan is to further consolidate / cleanup after this initial but big PR lands.

**Review by neonichu — APPROVED** [review]

File: Sources/Commands/Describe.swift
Comment (abertelrud) [inline Sources/Commands/Describe.swift:114]: Should we do a follow-on to separate out the described location similarly?

File: Sources/PackageGraph/LocalPackageContainer.swift
Comment (abertelrud) [inline Sources/PackageGraph/LocalPackageContainer.swift:46]: Could we add the package identity to the message here (or is it perhaps already part of the metadata?).  That would make it much easier to try to figure out errors coming in from the field that have this message.

File: Sources/PackageGraph/LocalPackageContainer.swift
Comment (abertelrud) [inline Sources/PackageGraph/LocalPackageContainer.swift:97]: Same comment as the other message.  If it's already part of the context for whatever handler wraps this call, so it will show up as metadata for a diagnostic, then that's even better.

File: Sources/PackageModel/Manifest/PackageDependencyDescription.swift
Comment (abertelrud) [inline Sources/PackageModel/Manifest/PackageDependencyDescription.swift:43]: In the past I think we had some issues with Foundation representing `git@...` URLs, at least on Darwin.  I agree that using URL instead of String here makes more sense conceptually, but do we think there might be side effects?  We should of course report them if there, but this might be something to watch for.

File: Sources/PackageModel/Manifest/PackageDependencyDescription.swift
Comment (abertelrud) [inline Sources/PackageModel/Manifest/PackageDependencyDescription.swift:153]: This is for a Git repository stored at a `file://` path?

File: Sources/PackageModel/PackageReference.swift
Comment (abertelrud) [reply Sources/PackageModel/PackageReference.swift:70]: Great, this will do a lot to clarify the meaning of the code.

**Review by abertelrud — APPROVED** [review]

File: Sources/Commands/Describe.swift
Comment (tomerd) [reply Sources/Commands/Describe.swift:114]: yes we could, less critical at this point imo

File: Sources/PackageGraph/LocalPackageContainer.swift
Comment (tomerd) [reply Sources/PackageGraph/LocalPackageContainer.swift:46]: sure. this is an internal error so a programmer error and not user one, so less important IMO

File: Sources/PackageModel/Manifest/PackageDependencyDescription.swift
Comment (tomerd) [reply Sources/PackageModel/Manifest/PackageDependencyDescription.swift:43]: this PR helps remove the logic that depends on checking if there is a scheme or not, and instead uses the type: `localSourceControl` vs `remoteSourceControl`, so ideally this should no longer be an issue regardless

Comment (abertelrud) [reply Sources/PackageModel/Manifest/PackageDependencyDescription.swift:43]: I see.  Yes, if SwiftPM is no longer asking those same questions but mainly using the URL as a container of URL-formatter strings, then that might go better.

File: Sources/Commands/Describe.swift
Comment (abertelrud) [reply Sources/Commands/Describe.swift:114]: Yes, absolutely.

File: Sources/PackageModel/Manifest/PackageDependencyDescription.swift
Comment (tomerd) [reply Sources/PackageModel/Manifest/PackageDependencyDescription.swift:43]: yes this is the goal - make the concrete location opaque to logic, so it only gets used when we pass it to git, where we use absoluteString to write out the full value

is there a specific use case you recall that we need to add a test for to make sure this works correctly in that sense?

**Pull request merged**
